### PR TITLE
urlapi: cleanup scheme parsing

### PR DESCRIPTION
**Pull request by icing**
Just a suggestion coming into my mind while thinking about #8041. Feel free to ignore.

`urlapi.c: Curl_is_absolute_url()`

  * always have a defined return in `buf`(if given)
  * only copy when scheme is detected

### CODE REVIEW

**Comment by bagder** [comment]
It doesn't lowercase the string, which makes test 1560 fail: https://cirrus-ci.com/task/5019655648575488?logs=test#L2482

And when writing this, I also realize that the check on https://github.com/curl/curl/blob/18270893abdb19f0ca170c118f8a2847dbd304be/lib/urlapi.c#L820 should rather use plain strcmp...

**Comment by icing** [comment]
made the change.

File: lib/urlapi.c [hunk 1]
```diff
@@ -251,13 +251,15 @@ bool Curl_is_absolute_url(const char *url, char *buf, size_t buflen)
...
-        break;
+      break;
     }
   }
   if(i && (url[i] == ':') && (url[i + 1] == '/')) {
     if(buf) {
-      memcpy(buf, url, i);
       buf[i] = 0;
+      while(i--) {
+        buf[i] = TOLOWER(url[i]);
```
Comment (bagder) [inline lib/urlapi.c:261]: I think this needs the typecast from the original to avoid warnings:
~~~c
buf[i] = (char)TOLOWER(url[i]);

**Comment by bagder** [comment]
I think this identifies another bug in the previous code as well. The third argument to `Curl_is_absolute_url` is confusingly used both as buffer length as well as maximum length acceptable for a scheme. I think maybe we should use it only as a buffer length and *always* use `MAX_SCHEME_LEN` as the max scheme length?

Either way, I fixed test 1560 by adding this patch:
~~~diff
diff --git a/lib/urlapi.c b/lib/urlapi.c
index 9005f040b..f1ec4713d 100644
--- a/lib/urlapi.c
+++ b/lib/urlapi.c
@@ -816,11 +816,11 @@ static CURLUcode seturl(const char *url, CURLU *u, unsigned int flags)
     return CURLUE_OUT_OF_MEMORY;
 
   hostname = &path[urllen + 1];
   hostname[0] = 0;
 
-  if(Curl_is_absolute_url(url, schemebuf, sizeof(schemebuf))) {
+  if(Curl_is_absolute_url(url, schemebuf, sizeof(schemebuf) - 1)) {
     url_has_scheme = TRUE;
     schemelen = strlen(schemebuf);
   }
 
   /* handle the file: scheme */
@@ -1525,11 +1525,11 @@ CURLUcode curl_url_set(CURLU *u, CURLUPart what,
     CURLUcode result;
     char *oldurl;
     char *redired_url;
     CURLU *handle2;
 
-    if(Curl_is_absolute_url(part, NULL, MAX_SCHEME_LEN + 1)) {
+    if(Curl_is_absolute_url(part, NULL, MAX_SCHEME_LEN)) {
       handle2 = curl_url();
       if(!handle2)
         return CURLUE_OUT_OF_MEMORY;
       result = parseurl(part, handle2, flags);
       if(!result)

**Comment by bagder** [comment]
On second thoughts, maybe just apply my fix and then work on that little snag in a separate PR.

**Review by bagder — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**